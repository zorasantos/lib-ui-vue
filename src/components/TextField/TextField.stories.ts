import type { Meta, StoryObj } from "@storybook/vue3";

import TextField from "./TextField.vue";

const meta = {
  title: "TextField",
  component: TextField,
  tags: ["autodocs"],
  argTypes: {},
} satisfies Meta<typeof TextField>;

export default meta;

type Story = StoryObj<typeof TextField>;

export const Default: Story = {
  args: {
    label: "Name",
    id: "name",
    errorMessages: "",
  },
};

export const WithError: Story = {
  args: {
    label: "Name",
    id: "name",
    errorMessages: "Name is field required!",
  },
};
