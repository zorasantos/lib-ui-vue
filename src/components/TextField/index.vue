<script setup lang="ts">
import { ref } from "vue";

defineProps<{
  label?: string;
  id?: string;
  errorMessages?: string | undefined;
  modelValue?: string | number;
}>();

const inputFocus = ref(false);

const handleFocus = () => {
  inputFocus.value = true;
};

const handleBlur = () => {
  inputFocus.value = false;
};
</script>
<template>
  <div class="mb-8">
    <label :for="id" class="sr-only">{{ label }}</label>
    <div class="relative">
      <input
        v-bind="$attrs"
        :value="modelValue"
        @input="(event: Event) => $emit(`update:modelValue`, (event.target as HTMLInputElement).value)"
        @focus="handleFocus()"
        @blur="handleBlur()"
        :class="`h-14 bg-zinc-100 hover:bg-zinc-200 rounded-t-md block w-full p-2.5 ${
          !inputFocus && 'placeholder-transparent'
        } placeholder-gray-400 focus:outline-none border-b focus:bg-[#D6D6D6]`"
      />
      <label
        :for="id"
        :class="`absolute left-3 transition-all ${
          modelValue || inputFocus
            ? 'text-neutral-950 text-xs -translate-y-3 top-3'
            : 'text-base text-gray-500 -translate-y-1 top-[35%]'
        } ${errorMessages ? 'text-red-500' : 'text-gray-600'}`"
      >
        {{ label }}
      </label>
    </div>

    <span class="text-red-500 text-xs font-semibold" v-if="errorMessages">{{
      errorMessages
    }}</span>
  </div>
</template>
