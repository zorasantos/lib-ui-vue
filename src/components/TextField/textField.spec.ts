import { mount } from "@vue/test-utils";
import TextField from "./index.vue";

describe("TextField", () => {
  test("renders label and input correctly", () => {
    const wrapper = mount(TextField, {
      props: {
        label: "Name",
        id: "name",
        type: "text",
        errorMessages: undefined,
        modelValue: "Test Name",
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find("label").text()).toBe("Name");
    expect(wrapper.find("input").element.value).toBe("Test Name");
    expect(wrapper.find("label").classes("text-red-500")).toBe(false);
  });

  test("displays error message when errorMessages prop is provided", () => {
    const wrapper = mount(TextField, {
      props: {
        label: "Nome",
        id: "Name",
        type: "email",
        errorMessages: "Nome é um campo obrigatório!",
        modelValue: "",
      },
    });

    expect(wrapper.find("label").classes("text-red-500")).toBe(true);
    expect(wrapper.find("label").classes("text-gray-600")).toBe(false);

    expect(wrapper.text()).toContain("Nome é um campo obrigatório!");
    expect(wrapper.find("span").classes("text-red-500")).toBe(true);
  });

  test("emits input event with correct value when input is typed", async () => {
    const wrapper = mount(TextField, {
      props: {
        label: "Name",
        id: "name",
        type: "text",
        errorMessages: undefined,
        modelValue: "",
      },
    });

    await wrapper.find("input").setValue("John Doe");

    const value = wrapper.emitted("update:modelValue") as Array<[][]>;

    expect(wrapper.emitted("update:modelValue")).toHaveLength(1);
    expect(value[0][0]).toBe("John Doe");
  });
});
