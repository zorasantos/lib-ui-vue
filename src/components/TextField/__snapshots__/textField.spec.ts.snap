// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextField > renders label and input correctly 1`] = `
"<div class=\\"mb-8\\" type=\\"text\\"><label for=\\"name\\" class=\\"block mb-2 text-base font-semibold text-gray-600\\">Name</label><input type=\\"text\\" class=\\"border border-gray-600 sm:text-sm rounded-lg block w-full p-2.5 placeholder-gray-400 focus:outline-none focus:border-blue-500 focus:ring-1 focus:ring-blue-500\\">
  <!--v-if-->
</div>"
`;
